From 8b538cca258a0c305564ffc09954ea6a33db5be7 Mon Sep 17 00:00:00 2001
From: Roman Podoliaka <rpodolyaka@mirantis.com>
Date: Mon, 6 Apr 2015 17:53:51 +0300
Subject: [PATCH 22/34] Fix rebuild of an instance with a volume attached

When detaching block devices on rebuild we only notify Cinder it's
safe to detach a volume, but don't actually tell the driver to do
that first.

Closes-Bug: #1440762

Change-Id: I017bf749f426717dc76cf99a387102848fb1c541
(cherry picked from commit 337471bc71cbdabe6b492379c81470abac8040d2)
---
 nova/compute/manager.py            | 2 +-
 nova/tests/compute/test_compute.py | 8 +++++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/nova/compute/manager.py b/nova/compute/manager.py
index 60ecd81..e3deb57 100644
--- a/nova/compute/manager.py
+++ b/nova/compute/manager.py
@@ -2790,7 +2790,7 @@ class ComputeManager(manager.Manager):
             def detach_block_devices(context, bdms):
                 for bdm in bdms:
                     if bdm.is_volume:
-                        self.volume_api.detach(context, bdm.volume_id)
+                        self.detach_volume(context, bdm.volume_id, instance)
 
             files = self._decode_files(injected_files)
 
diff --git a/nova/tests/compute/test_compute.py b/nova/tests/compute/test_compute.py
index 47e1ac1..015b7d6 100644
--- a/nova/tests/compute/test_compute.py
+++ b/nova/tests/compute/test_compute.py
@@ -11108,7 +11108,8 @@ class EvacuateHostTestCase(BaseTestCase):
                   'source_type': 'volume',
                   'device_name': '/dev/vdc',
                   'delete_on_termination': False,
-                  'volume_id': 'fake_volume_id'}
+                  'volume_id': 'fake_volume_id',
+                  'connection_info': '{}'}
 
         db.block_device_mapping_create(self.context, values)
 
@@ -11123,6 +11124,11 @@ class EvacuateHostTestCase(BaseTestCase):
             result["detached"] = volume["id"] == 'fake_volume_id'
         self.stubs.Set(cinder.API, "detach", fake_detach)
 
+        self.mox.StubOutWithMock(self.compute, '_detach_volume')
+        self.compute._detach_volume(mox.IsA(self.context),
+                                    mox.IsA(instance_obj.Instance),
+                                    mox.IsA(objects.BlockDeviceMapping))
+
         def fake_terminate_connection(self, context, volume, connector):
             return {}
         self.stubs.Set(cinder.API, "terminate_connection",
-- 
2.4.3

